DOC: Add sphinx-last-updated-by-git extension

mgeier opened the pull request:
Currently, each page of the docs shows the same "last updated" date.
IMHO, that's of limited usefulness.

Using this extension (full disclosure: written by me) each page should get an individual date, based on the Git commit where it was last changed.

https://github.com/mgeier/sphinx-last-updated-by-git

I'm not sure though, how well that works on such a huge project, because `git` is called at least once for each page. I've never tried it on something of similar scope.

Sorry for using NumPy as guinea pig ...

mgeier committed:
DOC: Add sphinx-last-updated-by-git extension

mgeier commented:
OK, sadly this doesn't seem to be working very well.
The Sphinx build on CircleCI timed out.

I guess I'll have to find a way to make fewer `git` calls.

For now, I'll close this PR.

eric-wieser commented:
This extension looks like a great idea. I wonder if by using GitPython you can avoid some overhead from repeated subprocess calls - I get the impression it might work directly with the git filesystem and/or cache commit information, although I may be mistaken.

mgeier commented:
Thanks for the suggestion!

AFAICT, GitPython also simply calls `git`:

https://github.com/gitpython-developers/GitPython/blob/3c19a6e1004bb8c116bfc7823477118490a2eef6/git/cmd.py#L173

https://github.com/gitpython-developers/GitPython/blob/3c19a6e1004bb8c116bfc7823477118490a2eef6/git/cmd.py#L998

https://github.com/gitpython-developers/GitPython/blob/3c19a6e1004bb8c116bfc7823477118490a2eef6/git/cmd.py#L724-L736

eric-wieser commented:
my understanding was that most of `GitPython` does not use the `cmd` module. There seems to be a lot of code that walks the git filesystem in other files in that repo.

mgeier commented:
Thanks for the hint!

I think I would want to use something like [Repo.iter_commits()](https://gitpython.readthedocs.io/en/stable/reference.html#git.repo.base.Repo.iter_commits) ...?

But this seems to call `repo.git.rev_list(rev, args, as_process=True, **kwargs)` (via `Commit.iter_items()`), which itself runs a new `subprocess`.

It looks like they obtain a list of SHAs from the `git` command, but then "parse the actual commit information directly from our lighting fast object database", see https://github.com/gitpython-developers/GitPython/blob/0c6e67013fd22840d6cd6cb1a22fcf52eecab530/git/objects/commit.py#L248-L277

AFAIU, this would still need a call to `git` for each file ...

I think a possible approach would be to collect a whole list of files in the Sphinx extension and then call `git log` once to get a customized list of commits and then manually parse the "last changed" information from this list of commits for the whole list of files in one go.

mgeier commented:
In the meantime I've created a PR with some refactoring in order to reduce the `git` calls: https://github.com/mgeier/sphinx-last-updated-by-git/pull/25

I've re-opened this PR as WIP and added my refactored branch: ab712cc54c1491baafd6c2d2257be8e5d62d4705

Sadly, this still times out the build process: https://app.circleci.com/pipelines/github/numpy/numpy/6016/workflows/4d82d6f9-2bd8-4268-9486-46cb1eb7ba52/jobs/18107/steps

Any ideas how I could improve this?

I guess I'll have to display some timing information to see where most of the time is spent ...

mattip commented:
All the pages in `generated` are built from the numpy docstrings, so you could skip those.

mgeier commented:
OK, I did a bit of debugging and it turns out it was simply a bug (or rather two bugs) that caused my code to be stuck, therefore causing a build timeout.

The `git log` output for merge commits was different than I expected (it didn't contain the added/changed file names), so parsing of the log went wrong.

Now I've added the `--first-parent` option (https://github.com/mgeier/sphinx-last-updated-by-git/pull/25/commits/ca08b58a28d5f4664469d03ae3c06eb4aecfec9b) which does include the file names as expected.

This means that the "last changed" date will actually be the date of the merge commit and not the individual commit within the merged branch. In other words, it will be the "change accepted date" rather than the "change authored date".

I hope this is not a dealbreaker!

If you think using the "change authored date" would be much better, we would have to find appropriate `git log` options that allow reliably parsing this information.

The second bug was much harder to find, because it didn't happen locally on my computer.
It turns out that CircleCI runs git 2.20.1, which seems to have different default behavior to my local git 2.30.0!
I found out (more or less by accident) that adding `-m` to the options makes the old version do the same as the new version (and I think there the flag has no effect).
See also https://github.com/mgeier/sphinx-last-updated-by-git/pull/25/commits/87685c65bc1561d56baaf31988ea811526d48b18.

In case future git versions change again (or someone uses an even older git version), I've added some `assert` statements that now hopefully raise an error instead of blocking further execution.

And now, there is finally a rendered result (with different "last updated" info on each page!): https://18137-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

> All the pages in `generated` are built from the numpy docstrings, so you could skip those.

That would be an option, but now it turns out that the `git` calls only take a few seconds all together, so this will not be necessary.

BTW, these auto-generated files get no "last updated" time anyway (as long as they don't have any dependencies known by Sphinx, which are in turn under version control).

In the case of auto-generated files, `git ls-files` is called once per directory to check whether files are controlled by Git or not. This takes about 15 to 20 milliseconds per directory.

eric-wieser commented:
First-parent sounds like the right choice to me.

mgeier commented:
> First-parent sounds like the right choice to me.

Probably, I'm not sure.

I think arguments can be made for and against it.

Initially I liked the idea of using the "PR accepted date", but I see two limitations with that:

* if a project uses a `git rebase` workflow, the merge date will not be visible (but we could change from "author date" to "commiter date", then it would be fine). The date of fast-forward merges would obviously not be visible, because they don't change anything on the commits. In other words, the date is inconsistent between different merging strategies (which may be used within one project).
* When looking at a source file on Github, it often shows a different date under "latest commit", e.g.
  * Oct 27, 2020: https://18137-908607-gh.circle-artifacts.com/0/doc/build/html/user/basics.rec.html
  * Oct 26, 2020: https://github.com/numpy/numpy/blob/master/doc/source/user/basics.rec.rst

For these reasons I think it would be better to *not* use `--first-parent`.

I initially used `--first-parent` because it was the only setting that created properly parse-able output, but in the meantime I found out that the `-m` flag (which I needed for old `git` versions anyway) actually leads to the output I was initially expecting.

@eric-wieser What do you think?

I now get the log output I want, and the "build devdocs w/ref warnings" finishes successfully (now even with status output from my extension!), but the "build devdocs" step times out.

I'm not sure whether that's caused by my extension, because the generated HTML artifact files (even though the CSS seems to be missing) *do* contain different dates on each one.

mgeier commented:
> the "build devdocs" step times out.

I have re-based this PR and now CircleCI succeeds, hooray!

Rendered: https://18162-908607-gh.circle-artifacts.com/0/doc/build/html/index.html

I guess the remaining CI failures are unrelated, right?

rgommers commented:
I like the end goal here. But I'm not sure it is doable - the docs need to build also when git is not available. There's at least two scenarios I can think of where this happens:
- Building docs from a source release (sdist). This is the most important one.
- There can be cases, in particular on Windows with some terminals, where `git` may be available in the terminal but not when called via (e.g.) `subprocess.Popen`. This is a bit niche, but we've fixed bugs for such cases before.

mgeier commented:
@rgommers That's a good point!

How common are those situations?
What are acceptable failure modes for those situations?

For the case that `git` is not found, we could create a Sphinx warning instead of an exception.

Alternatively, we could provide an option (that can also be used on the command line) to specify the Git executable name. Specifying an empty string could disable all `git` calls (and show no "last updated" dates at all).

For the case that `git` is available but there is no Git repo, we could also create a Sphinx warning instead of an exception.

However, if there were a way to disable all `git` calls, we could keep the exception and just mention the option to disable Git in the exception message (which would be shown at the end of the failed Sphinx run, by default without traceback).

Any other suggestions?

In case of doubt, I prefer hard errors over warnings, because latter can easily be overlooked (especially in NumPy's CI). But if hard errors are unacceptable in some situations, I'm open to changing them into warnings.

rgommers commented:
> How common are those situations?
> What are acceptable failure modes for those situations?

They're not super common, but they can be important. For example, I believe to create a Debian package for the SciPy docs (it's a separate `python-scipy-doc` package with build html), the package maintainer will start from the official sdist on PyPI.

In general, the sdist should be a standard archive or snapsnot of development, and work as expected without requiring the full development history that's in a git clone.



> For the case that `git` is not found, we could create a Sphinx warning instead of an exception.

Makes sense, but only if it's a one-off. No "one warning per html page" please, because that would make a build log unreadable quickly. One warning is nice indeed, because we aim to check for warnings in CI and error out for `>0` warnings.


> Alternatively, we could provide an option (that can also be used on the command line) to specify the Git executable name. Specifying an empty string could disable all `git` calls (and show no "last updated" dates at all).

Detecting if git is not present, and then by default (i.e., without the user having to figure out what's wrong and specifying a command-line flag) inserting something like `"built <today's date>, last updated: unknown"` sounds like a good option.

mgeier commented:
I've changed the git errors to warnings.

And those warnings can be silenced (in `conf.py` or on the command line) if desired:

```python
suppress_warnings = ['git.command_not_found', 'git.subprocess_error']
```

All git-related warnings can be suppressed by simply using `'git'`.

> Makes sense, but only if it's a one-off.

Yes, if `git` is not available, this currently causes only one warning (and disables all "last updated" displays).

If, however, git is available but the Sphinx build is run on a non-git-aware checkout (e.g. `sdist`), this currently causes 19 warnings: one for each source directory.

I don't really want to limit this to one warning, because theoretically, in each source directory a different git error could happen (or none at all), and this information would not be accessible if there were only one warning.

But if undesired, those warnings can be silenced using the `git.subprocess_error` category.

> No "one warning per html page" please, because that would make a build log unreadable quickly.

One per page will definitely not happen, because I just refactored my extension to call `git` per directory instead of per file, which was much too slow for such a big repo like NumPy's.

> One warning is nice indeed, because we aim to check for warnings in CI and error out for `>0` warnings.

On CI this shouldn't be relevant, because `git` should be available and everything should happen in a Git repo, right?

If everything goes as expected on CI, there will be no warnings at all.

> Detecting if git is not present, and then by default (i.e., without the user having to figure out what's wrong and specifying a command-line flag) inserting something like `"built <today's date>, last updated: unknown"` sounds like a good option.

Are you sure the replacement text is necessary?

This would need additional changes in the HTML templates and probably also additional custom fields in the HTML `context` variable.

I don't think that's worth it.
This is only for the exceptional case where `git` cannot be used, and even then the build date isn't really a very useful piece of information, is it? And "last updated: unknown" is just noise.

rgommers commented:
> If, however, git is available but the Sphinx build is run on a non-git-aware checkout (e.g. `sdist`), this currently causes 19 warnings: one for each source directory.

That doesn't sound too bad, should be fine.

rgommers commented:
> Are you sure the replacement text is necessary?

I'm not attached to the exact form, as long as the end result is reasonable. My main point was not to fail and make the person trying to build the docs manually specify something before the doc build can succeed.

mgeier commented:
> My main point was not to fail and make the person trying to build the docs manually specify something before the doc build can succeed.

After my latest changes, the Sphinx build will succeed even without `git` or if `git` exits with an error.
Manual intervention is only necessary to suppress the warnings.

However, if there is a `git` error, the HTML pages will have no "Last updated on ..." line and except the lack of that line, nothing about the `git` error is visible in the HTML pages.

Is this reasonable?

rgommers commented:
That sounds good to me, thanks @mgeier

mgeier commented:
Thanks for all the feedback so far!

I've just made a new release of my extension and I'm marking this PR "ready for review".

mattip commented:
We discussed this in the latest documentation meeting. The general feeling was that it would be more distracting than informative. It also adds one more thing that can break to our already complicated doc build process. @mgeier would you like to come to the next documentation team meeting on March 1 and explain the value this adds?

mgeier commented:
>  @mgeier would you like to come to the next documentation team meeting on March 1 and explain the value this adds?

Sure, as long as it's at a reasonable hour. I'm at Central European Time.

melissawm commented:
> > @mgeier would you like to come to the next documentation team meeting on March 1 and explain the value this adds?
> 
> Sure, as long as it's at a reasonable hour. I'm at Central European Time.

@mgeier the meeting will be at 8AM Pacific Time, and details for joining will be sent to the mailing list at the end of this week. Let me know if you need any more information.

mgeier commented:
As just discussed on the NumPy docs meeting, I'm closing this and as an alternative I've created #18520, which should just remove the line with the date.
